### DISCUSSION TITLE
How can we do modal window in egui?

### DISCUSSION THREAD

**Discussion by jackyzsn** [emilk/egui#1740] [Q&A]
How can we do modal window in egui? I checked the popup is close but not what I want. Demo app doesn't have something similar. Thanks.

**Comment by jackyzsn** [comment]
End up using native-dialog in my app for showing modal dialog

**Reply by coderedart** [reply]
there's Rusty File Dialog which is popular

**Comment by Barugon** [comment]
I use a flag to disable the main UI when I show a window that I want to be modal. When the window is no longer shown then the UI is enabled again.

**Comment by mrkaragoz** [comment]
> I use a flag to disable the main UI when I show a window that I want to me modal. When the window is no longer shown then the UI is enabled again.

How did you do that? Can you post a simple example?

**Reply by Barugon** [reply]
> How did you do that? Can you post a simple example?

Something like this:

```Rust
use eframe::egui;

#[derive(Default)]
struct App {
    window_open: bool,
}

impl eframe::App for App {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        if self.window_open {
            egui::Window::new("Modal Window")
                .open(&mut self.window_open)
                .show(ctx, |ui| {
                    ui.label("contents");
                });
        }

        egui::CentralPanel::default().show(ctx, |ui| {
            ui.set_enabled(!self.window_open);
            if ui.button("Open the window").clicked() {
                self.window_open = true;
            }
        });
    }
}

fn main() {
    eframe::run_native(
        "Modal window",
        eframe::NativeOptions::default(),
        Box::new(|_cc| Box::new(App::default())),
    );
}
```

**Comment by jackyzsn** [comment]
I'm using another dependency.  

[dependencies]
 native-dialog = "0.6.3"


                            MessageDialog::new()
                                .set_type(MessageType::Info)
                                .set_title("Success")
                                .set_text("Connected ..")
                                .show_alert()
                                .unwrap();

**Reply by v-maxson** [reply]
This seems to be the correct answer (at least for me). Seems to be *mostly* cross-platform which is good!

**Reply by Barugon** [reply]
Unfortunately, native-dialog (and rfg) manifest bad behavior on Gnome. The dialogs are not truly modal, so they can be behind your app window. Also, you have to put them in a separate thread (which may be problematic for Mac) or the system will think your app is not responding.

**Reply by John-Nagle** [reply]
> The dialogs are not truly modal, so they can be behind your app window.

Yes. This is a problem with file pickers. Today, you have to use the system's file picker, because file pickers are privileged. In various sandbox modes, they grant the program access to the file. How this works varies with the platform. But you can't just code a file picker in egui.
```
    let task = rfd::AsyncFileDialog::new()
        .set_title(t!("title.open_replay", state.get_lang()))
        .add_filter("json", &["json"])
        .set_parent(window)
        .pick_file();
```
The `set_parent`, given the parent window's `&winit::window::Window`, tells the file dialog that its parent is the main window. This does exactly the right thing on Windows - the file dialog is always in front, and clicking on the main window will not change that. It just gets you a beep.

Linux, (X/Gnome, at least), totally ignores that parenting. I brought this up for the **rfd** crate, and it's being worked on upstream. If you click on the main window, the main window gains focus, the file dialog goes behind it, and, in full screen mode, the user is very frustrated, since they have no easy way to get the file dialog on top again.

Anyone tried Mac?

If you want to try this, see my [ui-mock](https://github.com/John-Nagle/ui-mock), which is the GUI for a game minus the game. You can run it and try egui menus, which aren't connected to anything. I'd appreciate someone building this in Mac and reporting back. Clicking on the big "Open replay file" will get you a file picker. It wants a .json file, but doesn't actually read it.

I wrote that thing to wring out all the common UI cases for a game (full screen, file picker, Rend3 running 3D, etc.). So try it and see if it breaks. You'll have to build it, with "cargo build --examples --release".

Except for file pickers, I'd stay within egui. Don't bring up a system dialog. You can make an egui dialog that acts modal, with the background window disabled, as others have pointed out.

**Reply by Barugon** [reply]
> Today, you have to use the system's file picker, because file pickers are privileged. In various sandbox modes, they grant the program access to the file. How this works varies with the platform. But you can't just code a file picker in egui.

https://github.com/Barugon/egui_file

I can't tell you if it will work on WASM but it should work for any other platform.

**Reply by jackyzsn** [reply]
I tested native-dialog in Mac. Mac behaviour is same as windows. If you click main window, it beeps but the dialog still shows in the front.

**Reply by John-Nagle** [reply]
> I tested native-dialog in Mac. Mac behaviour is same as windows. If you click main window, it beeps but the dialog still shows in the front.

Excellent. 

On Linux, the **rfd** people are struggling with this. There's an upstream problem. See https://github.com/PolyMeilex/rfd/issues/46

**Comment by drew-abbo** [comment]
If anyone stumbles across this now, [`egui::Modal`](https://doc.servo.org/egui/containers/modal/struct.Modal.html) is probably the solution you're looking for.

**Reply by GreatAttractor** [reply]
Are there already facilities in `egui` itself to show a resizable, movable `Window` as a modal? (I.e., without resorting to custom flags as in Barugon's example.)
